### PR TITLE
Adds a working version of the format (%) method on strings

### PR DESCRIPTION
**Pull request by zamith**
There is a lot more it can do, but just this simple implementation is
helpful enough, I think. If you wish to learn more about this method,
you can read the [ruby documentation](http://ruby-doc.org/core-2.2.0/String.html#method-i-25).

### CODE REVIEW

**Comment by jhass** [comment]
In Ruby, `printf`, `sprintf` and `Stirng#%` are interfaces to the same code. We already got `sprintf` which uses `String::Formatter`, I think it would to enhance the latter to support hash substitutions, so we can keep it consistent too.

**Comment by zamith** [comment]
Sure. I did not know about that. Thanks.

**Comment by zamith** [comment]
@jhass Moved the functionality to the formatter. I don't like how the code turned out, and would love some feedback.

File: src/string/formatter.cr [hunk 1]
```diff
@@ -22,11 +22,33 @@ struct String::Formatter
...
+        break
+      else
+        word_chars << current_char
+      end
+      next_char
+    end
+    if current_arg.is_a?(Hash)
+      translations = current_arg as Hash(String, String)
+      @io << translations[word_chars.join]
+    end
```
Comment (jhass) [inline src/string/formatter.cr]: How about

```cr
key = String.build do |io|
  loop do
    case current_char
    when '}'
      break
    else
      io << current_char
    end
    next_char
  end
end

File: spec/std/string_spec.cr [hunk 2]
```diff
@@ -1306,4 +1306,16 @@ describe "String" do
...
+    it "substitutes one placeholder" do
+      res = "change %{this}" % { "this": "nothing" }
+      res.should eq "change nothing"
+    end
+
+    it "substitutes multiple placeholder" do
+      res = "change %{this} and %{more}" % { "this": "nothing", "more": "something" }
+      res.should eq "change nothing and something"
+    end
+  end
```
Comment (jhass) [inline spec/std/string_spec.cr]: I think a spec for when no hash is passed and one for when a hash is passed but the key is missing would be good.

File: src/string/formatter.cr (hunk 1 above)
Comment (zamith) [inline src/string/formatter.cr]: Definitely better. Is there no way around the type checking and casting? At least one that looks nicer?

File: spec/std/string_spec.cr (hunk 2 above)
Comment (zamith) [inline spec/std/string_spec.cr]: What do you mean by "no hash is passed"? Something like `"change %{this}" %`? It seems a bit of an odd situation. How should the language behave?

Comment (jhass) [inline spec/std/string_spec.cr]: I mean like `"%s %{foo}" % "bar"` and it should probably just ignore the substitutions like your current code does.

File: src/string/formatter.cr (hunk 1 above)
Comment (jhass) [inline src/string/formatter.cr]: I don't think so, if you look at the other methods in the class, you'll see they do about the same. Speaking of that, might be worth to extract the replacement into `substitution` or something to stay in style with the rest of the class.

Comment (asterite) [inline src/string/formatter.cr]: You can do:

```crystal
    if (arg = current_arg).is_a?(Hash)
      @io << arg[key]
    end
```

`current_arg` is a method call, so the compiler can't rely on it not changing type between calls.

Comment (zamith) [inline src/string/formatter.cr]: Cool. Didn't know about that. :)

File: spec/std/string_spec.cr (hunk 2 above)
Comment (zamith) [inline spec/std/string_spec.cr]: Hum... Isn't that just bad usage of the API? Do we usually have tests for that?

I think leaving them untouched is the best way.

Comment (zamith) [inline spec/std/string_spec.cr]: Really just wondering what is the norm, since in this case I think it makes sense to have that test, which I just added.

**Comment by jhass** [comment]
LGTM, though I don't think we need 7 commits for it ;) You can squash them using an interactive rebase (assuming you cloned from your fork):

```
git remote add upsstream https://github.com/manastech/crystal.git
git fetch upstream
git rebase -i origin/master # change pick to squash or fix, drop or reorder as you wish
git push -f orgin string_format # we edited the commits, so a force push is needed
```

Also keep in mind that you can edit a commit with `git commit --amend`.

**Comment by zamith** [comment]
Sure, I can squash it, np.

**Comment by asterite** [comment]
@zamith @jhass Thank you!!

I'll merge this, but I will change it so that:

```crystal
"change %{this}" % "this" #=> ArgumentError: one hash required
```

like in Ruby. And also fix this:

```crystal
"look: %{foo" % {"foo" => 1} # IndexError
```

but in Ruby it says "ArgumentError: malformed name - unmatched parenthesis".

**Pull request merged**